File: backend/scripts/migrate-to-mongo.ts
import { Client } from 'pg';
import { MongoClient, Decimal128 } from 'mongodb';
import dotenv from 'dotenv';
import path from 'path';

dotenv.config();

const PG_URL = process.env.DATABASE_URL || 'postgresql://charronix_user:charronix_pass_2024@localhost:5432/charronix_db';
const MONGO_URL = 'mongodb+srv://charlesjehanabad12345_db_user:QnxvTS546eTPe22a@cluster0.tui3r3t.mongodb.net/?appName=Cluster0';
const MONGO_DB_NAME = 'charronix_db';

async function migrate() {
    console.log('🚀 Starting Database Migration...');
    
    const pgClient = new Client({ connectionString: PG_URL });
    const mongoClient = new MongoClient(MONGO_URL);

    try {
        await pgClient.connect();
        await mongoClient.connect();
        console.log('✅ Connected to PostgreSQL and MongoDB');

        const db = mongoClient.db(MONGO_DB_NAME);

        // 1. Get all public tables
        const tableRes = await pgClient.query(`
            SELECT table_name 
            FROM information_schema.tables 
            WHERE table_schema = 'public' 
            AND table_type = 'BASE TABLE'
        `);
        const tables = tableRes.rows.map((r: any) => r.table_name);
        console.log(`📂 Found ${tables.length} tables to migrate: ${tables.join(', ')}`);

        for (const table of tables) {
            console.log(`\n⏳ Migrating table: ${table}...`);
            
            // 2. Fetch all data from the table
            const dataRes = await pgClient.query(`SELECT * FROM "${table}"`);
            const rows = dataRes.rows;

            if (rows.length === 0) {
                console.log(`⏩ Table ${table} is empty, skipping.`);
                continue;
            }

            // 3. Clean/Sanitize data for MongoDB
            const sanitizedRows = rows.map((row: any) => {
                const newRow: any = { ...row };
                for (const key in newRow) {
                    const val = newRow[key];
                    // Handle Decimal types
                    if (val && typeof val === 'object' && val.constructor.name === 'Decimal') {
                        newRow[key] = parseFloat(val.toString());
                    }
                    // Handle nulls if necessary (MongoDB stores nulls fine)
                }
                return newRow;
            });

            // 4. Insert into MongoDB
            const collection = db.collection(table);
            
            // Optional: Clear existing data in collection
            await collection.deleteMany({});
            
            const result = await collection.insertMany(sanitizedRows);
            console.log(`✅ Table ${table} migrated successfullly. Documents inserted: ${result.insertedCount}`);
        }

        console.log('\n✨ Migration Complete!');
        console.log(`🔗 You can now connect to MongoDB Compass using: ${MONGO_URL}${MONGO_DB_NAME}`);

    } catch (err) {
        console.error('❌ Migration Failed:', err);
    } finally {
        await pgClient.end();
        await mongoClient.close();
    }
}

migrate();

File: backend/scripts/migrate-to-mongo.py
import psycopg2
from psycopg2.extras import RealDictCursor
from pymongo import MongoClient
import os
import decimal
from datetime import datetime

PG_URL = 'postgresql://charronix_user:charronix_pass_2024@localhost:5432/charronix_db'
MONGO_URL = 'mongodb+srv://charlesjehanabad12345_db_user:QnxvTS546eTPe22a@cluster0.tui3r3t.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0'

print("🚀 Starting Database Migration (Python version to bypass Windows Node.js bug)...")

try:
    pg_conn = psycopg2.connect(PG_URL)
    mongo_client = MongoClient(MONGO_URL)

    db = mongo_client['charronix_db']

    with pg_conn.cursor() as cur:
        cur.execute("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' AND table_type = 'BASE TABLE'")
        tables = [row[0] for row in cur.fetchall()]

    print(f"📂 Found {len(tables)} tables to migrate.")

    for table in tables:
        print(f"⏳ Migrating table: {table}...")
        with pg_conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(f'SELECT * FROM "{table}"')
            rows = cur.fetchall()
            
        if not rows:
            print(f"⏩ Table {table} is empty, skipping.")
            continue
            
        # Sanitize types (Decimal, etc) for MongoDB
        for row in rows:
            for k, v in list(row.items()):
                if isinstance(v, decimal.Decimal):
                    row[k] = float(v)
                    
        collection = db[table]
        collection.delete_many({}) # Clear existing data
        collection.insert_many(rows)
        print(f"✅ Table {table} migrated successfully. Documents inserted: {len(rows)}")

    print(f"\n✨ Migration Complete! You can view it in MongoDB Compass with: {MONGO_URL}")
except Exception as e:
    print(f"❌ Migration Failed: {e}")
finally:
    if 'pg_conn' in locals():
        pg_conn.close()
    if 'mongo_client' in locals():
        mongo_client.close()
